spark/src/main/scala/org/apache/spark/sql/delta/commands/MergeIntoCommand.scala

```diff
@@ -249,14 +229,49 @@ case class MergeIntoCommand(
     //     target row is modified by multiple user or not
     // - the target file name the row is from to later identify the files touched by matched rows
     val joinType = if (notMatchedBySourceClauses.isEmpty) "inner" else "right_outer"
-    val targetDF = buildTargetPlanWithFiles(spark, deltaTxn, dataSkippedFiles)
+
+    // When they are only MATCHED clauses, we prune after the join the files that have no rows that
+    // satisfy any of the clause conditions.
+    val matchedPredicate =
+      if (isMatchedOnly) {
```

Comment:
A common scenario is:

```
MERGE INTO .... t
USING .... s
   ON s.id = t.id
   WHEN MATCHED AND NOT (s.name <=> t.name) THEN UPDATE SET t.name = s.name
   WHEN NOT MATCHED THEN INSERT (id, name) VALUES (s.id, s.name)
```

In that case the optimization won't work because it has `when not matched`?

Reply:
We can't skip data when reading the target in this case: let's assume there's a row `{"id": 10, "name": "Felipe"}` in both the source and target tables, i.e. it's a match but it won't be updated by the first `MATCHED` clause.

If we apply `NOT (s.name <=> t.name)` as a matched predicate here to prune files, then the file that contains that row in the target table may get pruned from the list of modified files. It won't be included in the join in `writeAllChanges` and the corresponding row in the source won't have a match anymore and will be wrongly inserted.

Reply:
Thanks for clarifying it. Do you know if in this case, we are re-writing the parquet file (with the same data, given the matched predicate is false).

I'm asking because user can have large tables and constantly update it using merge, and even if source and target are identical, we would rewrite all the data files. This is what I meant in #1812

Reply:
Yes, we would rewrite all files where at least one row matches the ON condition. I imagine that would be more complex to implement than the data skipping I'm adding here.